ADS-3831: Shopify market work around (#98)
Nosto/wiki-docs: features/shopify-markets-js-api.md

# Shopify Markets Workaround

{% hint style="info" %} Please note that this is a temporary workaround for loading recommendations based on Shopify Market until we release a complete end to end support. This workaround avoids any manual setup to be done on merchant's store. Instead, it involves minor changes to the existing recommendation templates.
{% endhint %}

## JS API to the rescue

### Drawbacks of current approach

> In order to workaround product pricing for Shopify Markets, the existing approach follows the changes outlined [here](/features/multi-currency-support.md#fetch-product-prices-in-presentment-currency-legacy). This involves updating Shopify theme liquid template manually in order to enable Shopify Markets. The approach is also limited to loading product prices for the market. 

Reviewer: Linking to this page might cause confusion. I think it would be clearer to provide the liquid snippet here on this page so everything the merchant needs to do is in one place.

Author: Not included those snippet inline here for the following reasons
1. It will duplicate the snippet
2. Merchants are already well aware of the page. So, though it shouldn't be much of hassle. Moreover, the actual snippet is not about SM workaround and need not be included inline here
Added the link  here so that the merchant will know the JS API is actually an alternative approach for the legacy pricing logic.

Reviewer: I think this page still needs to be clearer about the required theme changes.

1. Product URL - updates Product URL for every product in the recommendation to match the URL of the store's market. At the same time, the URL will retain all the attribution parameters from the original URL
2. Product Title - updates Product Title, for every product in the recommendation, to reflect the language of the current market
3. Product Description - updates Product description, for every product in the recommendation, to reflect the language of the current market
4. Product Category - updates Product category, for every product in the recommendation, to reflect the language of the current market
5. Product Variant Options - For those recommendations involving chosing variants, this API translates all the variant options and title to reflect the language of the current market

#### Usage
The API takes an Object `RecProductElementSelectors` as a single parameter. The following table lists all the allowed mandatory and optional parameters of `RecProductElementSelectors` Object

| Parameter Name | Required/Optional | Type | Description |
| :-------------------: | :---------------------: | :----------: | :-------------- |

> | productSectionElement | Required | String | CSS classname/Id selector of the `div` or an equivalent wrapper element that encapsulates all the product details and the variant options  |

Reviewer: ```suggestion
| productSectionElement | Required | String | CSS selector of the element containing all the product details and the variant options  |
```

"CSS classname/Id selector" is a bit confusing. It's simply a CSS selector.
"`div` or an equivalent wrapper element" It's a HTML element. It can be any type of element.

Author: Updated. Thanks!

> | productUrlElement |  Required  | String |CSS classname selector of all <a> elements/Id of the <a> element hosting the product URL |

Reviewer: ```suggestion
| productUrlElement |  Required  | String |CSS selector of all `a` elements within `productSectionElement` containing the product URL |
```

Author: Updated. Thanks!

> | productTitleElement | Required | String | CSS classname selecror/Id of the element hosting the product title |

Reviewer: ```suggestion
| productTitleElement | Required | String | CSS selector of all elements within the `productSectionElement` element containing the product title |
```

Author: Updated. Thanks!

> | productHandleAttribute | Required | String | Attribute name for fetching the product handle. This value can be fetched in recommendation template using `$!product.lastPathOfProductUrl`. This attribute has to be part of the `div` or an equivalent wrapper element hosting the `productSectionElement` |

Reviewer: ```suggestion
| productHandleAttribute | Required | String | Attribute name for fetching the product handle. This value can be fetched in recommendation template using `$!product.lastPathOfProductUrl`. This attribute has to be part of the same element referenced by `productSectionElement` |
```

Author: Updated. Thanks!

> | priceElement | Required | String | CSS classname selector of all elements or Id of the element displaying product `price` value. This parameter is required and is used to display the actual product price |

Reviewer: ```suggestion
| priceElement | Required | String | CSS selector of all elements within the `productSectionElement` element containing the product price |
```

Author: Updated. Thanks!

> | listPriceElement | Optional | String | CSS classname selector of all elements or Id of the element displaying product `compare_at_price` value. This is used only when the product has a discounted price and the recommendation displays both old & new price  |

Reviewer: ```suggestion
| listPriceElement | Optional | String | CSS selector of all elements within the `productSectionElement` element containing the product list price.  This is used only when the product has a discounted price and the recommendation displays both old & new price. |
```

Author: Updated. Thanks!

> | listPriceElement | Optional | String | CSS classname selector of all elements or Id of the element displaying product `compare_at_price` value. This is used only when the product has a discounted price and the recommendation displays both old & new price  |

Reviewer: This look duplicated with the line above.

Author: Removed!

> | defaultVariantIdAttribute | Optional | String | This parameter is used only when the recommendation doesn't contain variant options. When not provided, first variant of product is used. It is an Attribute element and is used for determining the product pricing for the market. It should be part of the `div` or an equivalent wrapper element hosting the `productSectionElement` |

Reviewer: ```suggestion
| defaultVariantIdAttribute | Optional | String | Attribute name for selecting the product variant. The first product variant is used if this is not provided. This attribute has to be part of the same element referenced by `productSectionElement`.
```

Author: Updated. Thanks!

> | descriptionElement | Optional | String | CSS classname selector of all elements or Id of the element displaying product description. Needed only when the recommendation displays product description |

Reviewer: ```suggestion
| descriptionElement | Optional | String | CSS selector of all elements within the `productSectionElement` element containing the product description |
```

Author: Updated. Thanks!

> | totalVariantOptions | Optional | Number | Total variant option levels displayed in the recommendation. Here variant level means, the selection of one variant value loads the immediate next variant and so on. More on this in the below sections. When not provided, actual product variant count is used. This value can be below the total actual product variants but not above that value. Currently, we support only a maxium of 5 variant option levels |

Reviewer: This description is quite difficult to read.

| position | Optional | Number | Must be provided when `optionElement` selector is provided. Numeric value of the order number/position of the variant option as declared in the Shopify admin product page. It's not required to display the elements, in the same order, in the recommendations, but it's requred to supply the appropriate position number and it should match the order in the Shopify admin |
| dependentVariantSelector | Optional | Object/VariantOptionSelector | Selectors configuration for sub-level variants that changes based on the value selected in this primary variant. A maximum of 5 sub-level variants can be supported |

#### Error reporting
In case of any issues with the API call or element structuring, the API will promptly report the errors in the browser console. Please contact Nosto support in case you come across these issues.

#### Limitations
1. Only allows upto 5 sub-level variants

> 2. Only allows 1 primary variants

Reviewer: ```suggestion
2. Only allows 1 primary variant.
```

Author: Updated. Thanks

> 3. Strict structuring necessary. In other words, it's necessary to follow the rules defined in the description of each API parameters.

Reviewer: Is this point necessary? I think it's pretty clear it won't work if you make a typo.

Author: It's not about type. Our logic expects some attributes to be part of the parent `productSectionElement` element. I included it here so that the merchants don't move them around.

Reviewer: So you mean the `productSectionElement` has some mandatory fields? In that case, isn't it better to improve validation using `validate.js`?

> 4. This is just a workaround until the actual realease of complete Shopify Markets support. 

Reviewer: This point was already made above.

Author: Removed

        <span class="nosto-newprice money">$!product.price</span>
    </a>
    <div class="nosto-rating">
        #set($ratingInt = $math.round($!product.ratingValue.getAsNumber()).intValue())
        #set($ratingRest = 5 - $ratingInt)
        #if($ratingInt > 0)
            #foreach ($number in [1..$ratingInt])
                <span class="nosto-star-full">★</span>
            #end
        #end

        #if($ratingRest != 0)
            #foreach ($number in [1..$ratingRest])
                <span class="nosto-star-empty">★</span>
            #end
        #end

        <span class="nosto-reviews">($!product.reviewCount)</span>
    </div>
    
    <div class="variant-options">
        <h3 class="size-option-title">Size</h3>
        <div class="nosto-sizes">
            #foreach($sku in $product.skus)
                <span x-nosto-product="$!product.productId" class="nosto-size #if(!$!sku.available) nosto-disabled #end">$!sku.getCustomField('Size')</span>
            #end
        </div>
        <h3 class="color-option-title">Color</h3>
        <div class="nosto-colors">
        </div>
    </div>
</div>
```

For the template above, you can call the JS API  

```javascript

> (function() {

Reviewer: None of this JS is related to the Markets workaround, is it? I think adding this code to the example could cause confusion.

Author: It's in fact about the JS workaround. The script provides an example of how the logic works when a value of variation dropdown changes.
